src/service/workspace_wrapper.py

```diff
@@ -124,9 +152,58 @@ async def check_workspace_permissions(self, workspace_ids: set[int]) -> None:
                     "Workspace.get_workspace_info", [{"id": wsi}], token=self._token
                 )
             except ServerError as e:
-                if any(x in e.message for x in ["may not read workspace", "No workspace with id"]):
-                    raise errors.DataPermissionError(
-                        f"The workspace service disallowed access to workspace {wsi}") from e
-                # Could be lots of other causes here, and the only way to tell is to inspect
-                # the workspace source code for error strings, so...
-                raise e
+                self._check_err(wsi, e)
+
+    def _check_err(self, wsid: int, e: ServerError):
+        if "has invalid reference" in e.message:
+            # error message is sort of gross but good enough for now, since this should never
+            # happen for collections, since the target upas are public and we're setting up
+            # the data types and don't expect type mismatch errors from the set @ws annotations
+            raise errors.DataPermissionError(e.message)
+        if any(x in e.message for x in 
+            ["may not read workspace", "No workspace with id", "may not write"]
+        ):
+            raise errors.DataPermissionError(
+                f"The workspace service disallowed access to workspace {wsid}") from e
+        # Could be lots of other causes here, and the only way to tell is to inspect
+        # the workspace source code for error strings, so...
+        raise e
+
+    def _obj_info_to_upa(self, objinfo) -> str:
+        return f"{objinfo[6]}/{objinfo[0]}/{objinfo[4]}"
+
+    async def save_sets(self, wsid: int, sets: list[SetSpec]) -> dict[str, str]:
+        """
+        Save one or more sets to the workspace.
+
+        wsid - the ID of the workspace to save to.
+        sets - the sets to save.
+
+        Returns a mapping of the resulting set UPAs to their type.
+        """
+        # should add some size checks in future, like max # of sets, max # of items in set, etc
+        if not sets:
+            raise ValueError("no sets")
+        objs = []
+        for s in sets:
+            if not s.upas:
+                raise ValueError("all sets must have at least one UPA")
+            setinfo = _TYPE_TO_SET_INFO.get(s.upa_type)
+            if not setinfo:
+                raise errors.IllegalParameterError(f"Unsupported workspace type: {s.upa_type}")
+            wsset = {"description": s.description} | setinfo["items"](s.upas)
```

Comment:
just want to make sure when s.description is none, ws.save also works for both sets type right? `description` is optional for both sets type?

Reply:
It's not optional, but the workspace allows `null` values for any primitive field.